airflow/timetables/events.py

```diff
@@ -0,0 +1,97 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Iterable, Optional
+
+import numpy as np
+import pendulum
+from pendulum import DateTime
+
+from airflow.timetables.base import DagRunInfo, DataInterval, TimeRestriction, Timetable
+
+
+class EventsTimetable(Timetable):
+    """
+    Timetable that schedules DAG runs at specific listed datetimes. Suitable for
+    predictable but truly irregular scheduling such as sporting events.
+
+    :param event_dates: List of datetimes for the DAG to run at
+    :param restrict_to_events: Whether manual runs should use the most recent event or
+    the current time
+    """
+
+    description = ""
+
+    def __init__(self, event_dates: Iterable[DateTime], restrict_to_events: bool = False, presorted=False):
+
+        self.event_dates = np.array(event_dates)
+        if not presorted:
+            # For long lists this could take a while, so only want to do it once
+            self.event_dates = np.sort(self.event_dates)
+        self.restrict_to_events = restrict_to_events
+        self.description = self.summary
+
+    @property
+    def summary(self) -> str:
+        return f"{len(self.event_dates)} Events"
+
+    def __repr__(self):
+        return self.summary
+
+    def next_dagrun_info(
+        self,
+        *,
+        last_automated_data_interval: Optional[DataInterval],
+        restriction: TimeRestriction,
+    ) -> Optional[DagRunInfo]:
+        if last_automated_data_interval is None:
+            next_event = self.event_dates.min()
+        else:
+            future_dates = self.event_dates[self.event_dates > last_automated_data_interval.end]
+            if len(future_dates) > 0:
+                next_event = future_dates.min()
+            else:
+                return None
```

Comment:
```suggestion
            future_dates = itertools.dropwhile(lambda when: when < last_automated_data_interval.end, self.event_dates)
            next_event = next(future_dates, None)
            if next_event is None:
                return None

Reply:
Neat! I will adapt this a couple of other places. Also it has to be `<=`. Thanks, not being able to figure this out is why I used numpy.